File: problem/Valid_Sudoku/isValidSudoku.py
class Solution:
    def isValidSudoku(self, board):
        rows = {}
        cols = {}
        boxes = {}
        
        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue
                
                if r not in rows:
                    rows[r] = set()
                if c not in cols:
                    cols[c] = set()
                if (r // 3, c // 3) not in boxes:
                    boxes[(r // 3, c // 3)] = set()
                
                if board[r][c] in rows[r] or board[r][c] in cols[c] or board[r][c] in boxes[(r // 3, c // 3)]:
                    return False
                
                rows[r].add(board[r][c])
                cols[c].add(board[r][c])
                boxes[(r // 3, c // 3)].add(board[r][c])
        
        return True

File: problem/Valid_Sudoku/isValidSudoku.c
bool isValidSudoku(char **board, int boardSize, int *boardColSize)
{
    bool rows[9][9] = {false};
    bool cols[9][9] = {false};
    bool boxes[9][9] = {false};

    for (int r = 0; r < 9; r++)
    {
        for (int c = 0; c < 9; c++)
        {
            if (board[r][c] == '.')
            {
                continue;
            }

            char value = board[r][c];
            int digit = value - '1';
            int boxIndex = (r / 3) * 3 + (c / 3);

            if (rows[r][digit] || cols[c][digit] || boxes[boxIndex][digit])
            {
                return false;
            }

            rows[r][digit] = true;
            cols[c][digit] = true;
            boxes[boxIndex][digit] = true;
        }
    }

    return true;
}
